File: java/src/test/java/com/taobao/top/link/endpoint/MessageIOTest.java
package com.taobao.top.link.endpoint;

import static org.junit.Assert.*;

import java.nio.ByteBuffer;
import java.util.Date;
import java.util.HashMap;

import org.junit.Test;

public class MessageIOTest {
	@Test
	public void read_write_test() {
		Message msg = new Message();
		msg.messageType = MessageType.SEND;
		msg.flag = 10;
		msg.token = "abc";
		msg.statusCode = 100;
		msg.statusPhase = "abcd";
		msg.content = new HashMap<String, Object>();
		msg.content.put("name1", "中文abc");
		msg.content.put("name2", "abc");

		msg.content.put("void", null);
		msg.content.put("byte", (byte) 1);
		msg.content.put("int16", (short) 100);
		msg.content.put("int32", (int) 100);
		msg.content.put("int64", (long) 100);
		msg.content.put("date", new Date());

		ByteBuffer buffer = ByteBuffer.allocate(1024);
		MessageIO.writeMessage(buffer, msg);
		// buffer.flip();//flip in write
		Message msg2 = MessageIO.readMessage(buffer);

		assertEquals(msg.messageType, msg2.messageType);
		assertEquals(msg.flag, msg2.flag);
		assertEquals(msg.token, msg2.token);
		assertEquals(msg.statusCode, msg2.statusCode);
		assertEquals(msg.statusPhase, msg2.statusPhase);
		assertEquals(msg.content.size(), msg2.content.size());
		assertEquals(msg.content.get("name1"), msg2.content.get("name1"));
		assertEquals(msg.content.get("name2"), msg2.content.get("name2"));

		assertEquals(msg.content.get("void"), msg2.content.get("void"));
		assertEquals(msg.content.get("byte"), msg2.content.get("byte"));
		assertEquals(msg.content.get("int16"), msg2.content.get("int16"));
		assertEquals(msg.content.get("int32"), msg2.content.get("int32"));
		assertEquals(msg.content.get("int64"), msg2.content.get("int64"));
		assertEquals(msg.content.get("date"), msg2.content.get("date"));
	}
}

File: csharp/src/Top.Link.Test/MessageIOTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using NUnit.Framework;
using Taobao.Top.Link.Endpoints;

namespace Taobao.Top.Link.Test
{
    [TestFixture]
    public class MessageIOTest
    {
        [TestCase]
        public void TypeEquals()
        {
            Assert.AreEqual(typeof(byte), typeof(Byte));
            Assert.AreEqual(typeof(string), typeof(String));
            Assert.AreEqual(typeof(short), typeof(Int16));
            Assert.AreEqual(typeof(int), typeof(Int32));
            Assert.AreEqual(typeof(long), typeof(Int64));
        }

        [TestCase]
        public void ReadWriteTest()
        {
            var s = new MemoryStream();
            var msg = new Message();
            msg.MessageType = MessageType.SEND;
            msg.StatusPhase = "error";
            msg.Content = new Dictionary<string, object>();
            msg.Content.Add("byte", (byte)1);
            msg.Content.Add("string", "string");
            msg.Content.Add("int16", (short)123);
            msg.Content.Add("int32", 156);
            msg.Content.Add("int64", 178L);
            msg.Content.Add("date", DateTime.Now);
            MessageIO.WriteMessage(s, msg);

            s.Position = 0;

            var msg2 = MessageIO.ReadMessage(s);
            Assert.AreEqual(msg.MessageType, msg2.MessageType);
            Assert.AreEqual(msg.StatusPhase, msg2.StatusPhase);
            foreach (var i in msg.Content)
            {
                Assert.AreEqual(msg.Content[i.Key], msg2.Content[i.Key]);
                Assert.AreEqual(msg.Content[i.Key].GetType(), msg2.Content[i.Key].GetType());
                Console.WriteLine(i.Key + "=" + msg2.Content[i.Key]);
            }
        }
    }
}